fdbserver/KeyValueStoreRocksDB.actor.cpp

```diff
@@ -43,127 +44,158 @@ rocksdb::Options getOptions() {
-				a.done.sendError(statusToError(s));
-			} else {
-				a.done.send(Void());
-			}
+	void action(OpenAction& a) {
+		std::vector<rocksdb::ColumnFamilyDescriptor> defaultCF = { rocksdb::ColumnFamilyDescriptor{
+			"default", getCFOptions() } };
+		std::vector<rocksdb::ColumnFamilyHandle*> handle;
+		rocksdb::DB * newDB = nullptr;
+		auto status = rocksdb::DB::Open(getOptions(), a.path, defaultCF, &handle, &newDB);
+		db.reset(newDB);
+		if (!status.ok()) {
+			handle.clear();
+			TraceEvent(SevWarn, "RocksDBError").detail("Warn", status.ToString()).detail("Method", "Open (will retry)");
+			// TODO: Back the database up instead of just nuking it!
+			rocksdb::DestroyDB(a.path, getOptions(), defaultCF);
+			status = rocksdb::DB::Open(getOptions(), a.path, defaultCF, &handle, &newDB);
+			db.reset(newDB);
+		}
+		if (!status.ok()) {
+			TraceEvent(SevError, "RocksDBError").detail("Error", status.ToString()).detail("Method", "Open");
+			a.done.sendError(statusToError(status));
+		} else {
+			a.done.send(Void());
 		}
+	}
 
-		struct CloseAction : TypedAction<Writer, CloseAction> {
-			ThreadReturnPromise<Void> done;
-			std::string path;
-			bool deleteOnClose;
-			CloseAction(std::string path, bool deleteOnClose) : path(path), deleteOnClose(deleteOnClose) {}
-			double getTimeEstimate() override { return SERVER_KNOBS->COMMIT_TIME_ESTIMATE; }
-		};
-		void action(CloseAction& a) {
+	struct CommitAction : TypedAction<Writer, CommitAction> {
+		std::unique_ptr<rocksdb::WriteBatch> batchToCommit;
+		ThreadReturnPromise<Void> done;
+		double getTimeEstimate() override { return SERVER_KNOBS->COMMIT_TIME_ESTIMATE; }
+	};
+	void action(CommitAction& a) {
+		rocksdb::WriteOptions options;
+		options.sync = true;
+		auto s = db->Write(options, a.batchToCommit.get());
+		if (!s.ok()) {
+			TraceEvent(SevError, "RocksDBError").detail("Error", s.ToString()).detail("Method", "Commit");
+			a.done.sendError(statusToError(s));
+		} else {
+			a.done.send(Void());
+		}
+	}
+
+	struct CloseAction : TypedAction<Writer, CloseAction> {
+		ThreadReturnPromise<Void> done;
+		std::string path;
+		bool deleteOnClose;
+		CloseAction(std::string path, bool deleteOnClose) : path(path), deleteOnClose(deleteOnClose) {}
+		double getTimeEstimate() override { return SERVER_KNOBS->COMMIT_TIME_ESTIMATE; }
+	};
+	void action(CloseAction& a) {
+		if (db) {
 			auto s = db->Close();
 			if (!s.ok()) {
 				TraceEvent(SevError, "RocksDBError").detail("Error", s.ToString()).detail("Method", "Close");
 			}
-			if (a.deleteOnClose) {
-				std::vector<rocksdb::ColumnFamilyDescriptor> defaultCF = { rocksdb::ColumnFamilyDescriptor{
-					"default", getCFOptions() } };
-				rocksdb::DestroyDB(a.path, getOptions(), defaultCF);
-			}
-			a.done.send(Void());
+			db = nullptr;
 		}
-	};
-
-	struct Reader : IThreadPoolReceiver {
-		DB& db;
-
-		explicit Reader(DB& db) : db(db) {}
-
-		void init() override {}
+		if (a.deleteOnClose) {
```

Comment:
I think the change in indentation tricked git.  This stanza should be inside the `if(db) { ... ` block immediately above.